File: TypeScript/src/Day10.ts
import Day from "./Day";
import * as Util from "./Util";


class Day10 implements Day<number[]>{
    parseInput(text: string): number[] {
        return text.split(/\r?\n/)
            .map((s: string) => parseInt(s));
    }

    solvePart1(input: number[]): string {
        let sortedInput: number[] = input.sort((m: number, n: number) => m - n);
        let joltageGapDistribution: Map<number,number> = this.joltageGapDistribution(sortedInput);
        let oneCount = Util.zeroIfInvalid(joltageGapDistribution.get(1));
        let threeCount = Util.zeroIfInvalid(joltageGapDistribution.get(3));
        return (oneCount * threeCount).toString();
    }

    private joltageGapDistribution(sortedAdapters: number[]): Map<number,number>{
        let distribution: Map<number,number> = new Map<number,number>();
        for(let index = 1; index < sortedAdapters.length; index++){
            let joltageGap: number = sortedAdapters[index] - sortedAdapters[index - 1];
            Util.addCount(distribution, joltageGap, 1);
        }
        Util.addCount(distribution, sortedAdapters[0], 1);
        Util.addCount(distribution, 3, 1);
        return distribution;
    }



    solvePart2(input: number[]): string {
        let sortedInput: number[] = input.sort((m: number, n: number) => m - n);
        let numberOfPossibilities = this.numberOfPossibleAdapterCombinations(sortedInput);
        return numberOfPossibilities.toString();
    }

    private numberOfPossibleAdapterCombinations(sortedAdapters: number[]) : number{
        if(sortedAdapters.length <= 0){
            return 0;
        }

        let combinationsByPreviousGap: [number, number, number] = [1, 1, 1];
        let index: number = sortedAdapters.length - 2;
        let currentGap: number;
        while(index >= 0){
            currentGap = sortedAdapters[index + 1] - sortedAdapters[index];
            combinationsByPreviousGap = this.combinationsByPreviousGap(combinationsByPreviousGap, currentGap);
            index--;
        }
        return combinationsByPreviousGap[sortedAdapters[0]-1];
    }

    private combinationsByPreviousGap(priorCombinationsByPreviousGap: [number, number, number], currentGap: number): [number, number, number]{
        switch(currentGap){
            case 1:
                return [
                    priorCombinationsByPreviousGap[0] + priorCombinationsByPreviousGap[1],
                    priorCombinationsByPreviousGap[0] + priorCombinationsByPreviousGap[2],
                    priorCombinationsByPreviousGap[0]
                ];
            case 2:
                return [
                    priorCombinationsByPreviousGap[1] + priorCombinationsByPreviousGap[2],
                    priorCombinationsByPreviousGap[1],
                    priorCombinationsByPreviousGap[1]
                ]
            case 3:
                return [
                    priorCombinationsByPreviousGap[2],
                    priorCombinationsByPreviousGap[2],
                    priorCombinationsByPreviousGap[2]
                ]
            default:
                return [0,0,0]
        }
    }
}


export default Day10;

File: Rust/adventOfCode2020/src/day/day10.rs


pub struct Day10 {}

impl super::Day for Day10{
    type PuzzleInput = Vec<usize>;

    fn parse_input(&self, text: std::string::String) -> Self::PuzzleInput {
        text.lines()
            .map(|line| line.parse::<usize>())
            .filter_map(Result::ok)
            .collect()
    }

    fn solve_part1(&self, mut adapters: Self::PuzzleInput) -> std::string::String {
        adapters.sort_unstable();
        let mut joltage_gaps = [0, 0, 0];
        joltage_gaps[adapters[0]-1] += 1;
        for pair in adapters.windows(2){
            let joltage_gap = pair[1] - pair[0];
            joltage_gaps[joltage_gap - 1] += 1;
        }
        joltage_gaps[3-1] += 1;
        let result = joltage_gaps[0] * joltage_gaps[3-1];
        result.to_string()
    }

    fn solve_part2(&self, mut adapters: Self::PuzzleInput) -> std::string::String {
        adapters.sort_unstable();
        let result = number_of_possible_adapter_combinations(&adapters);
        result.to_string()
    }
}

fn number_of_possible_adapter_combinations(sorted_adapters: &[usize]) -> usize{
    if sorted_adapters.is_empty(){
        return 1;
    }

    let mut combinations_by_gap = [1, 1, 1];
    for adapter_pair in sorted_adapters.windows(2).rev(){
        let current_gap = adapter_pair[1] - adapter_pair[0];
        combinations_by_gap = combinations_by_previous_gap(combinations_by_gap, current_gap);
    }

    combinations_by_gap[sorted_adapters[0] - 1]
}

fn combinations_by_previous_gap(prior_combinations_by_previous_gap: [usize; 3], current_gap: usize) -> [usize; 3]{
    match current_gap{
        1 => [
            prior_combinations_by_previous_gap[0] + prior_combinations_by_previous_gap[1],
            prior_combinations_by_previous_gap[0] + prior_combinations_by_previous_gap[2],
            prior_combinations_by_previous_gap[0],
        ],
        2 => [
            prior_combinations_by_previous_gap[1] + prior_combinations_by_previous_gap[2],
            prior_combinations_by_previous_gap[1],
            prior_combinations_by_previous_gap[1],
        ],
        3 => [
            prior_combinations_by_previous_gap[2],
            prior_combinations_by_previous_gap[2],
            prior_combinations_by_previous_gap[2],
        ],
        _ => [0, 0, 0],
    }
}






#[cfg(test)]
mod day10_tests {
    use super::*;
    use crate::input;
    use crate::day;

    fn example_input() -> String{
        String::from(
"16
10
15
5
1
11
7
19
6
12
4")
    }

    fn example2_input() -> String{
        String::from(
"28
33
18
42
31
14
46
20
48
47
24
23
49
45
19
38
39
11
1
32
25
35
8
17
7
9
4
2
34
10
3")
    }

    #[test]
    fn example_part1() {
        let day: Box<dyn day::DaySolver> = Box::new(Day10{});
        let problem_input = example_input();
        let expected_result = (7 * 5).to_string();
        let actual_result = day.solve_part1(problem_input);
        assert_eq!(actual_result, expected_result);
    }

    #[test]
    fn example2_part1() {
        let day: Box<dyn day::DaySolver> = Box::new(Day10{});
        let problem_input = example2_input();
        let expected_result = (22 * 10).to_string();
        let actual_result = day.solve_part1(problem_input);
        assert_eq!(actual_result, expected_result);
    }

    #[test]
    fn example_part2() {
        let day: Box<dyn day::DaySolver> = Box::new(Day10{});
        let problem_input = example_input();
        let expected_result = 8.to_string();
        let actual_result = day.solve_part2(problem_input);
        assert_eq!(actual_result, expected_result);
    }

    #[test]
    fn example2_part2() {
        let day: Box<dyn day::DaySolver> = Box::new(Day10{});
        let problem_input = example2_input();
        let expected_result = 19208.to_string();
        let actual_result = day.solve_part2(problem_input);
        assert_eq!(actual_result, expected_result);
    }

    #[test]
    fn correct_part1() {
        let day: Box<dyn day::DaySolver> = Box::new(Day10{});
        let problem_input = input::puzzle_input(&input::PuzzleConfiguration{day: 10, part: 1}).unwrap();
        let expected_result = String::from("2368");
        let actual_result = day.solve_part1(problem_input);
        assert_eq!(actual_result, expected_result);
    }

    #[test]
    fn correct_part2() {
        let day: Box<dyn day::DaySolver> = Box::new(Day10{});
        let problem_input = input::puzzle_input(&input::PuzzleConfiguration{day: 10, part: 2}).unwrap();
        let expected_result = String::from("1727094849536");
        let actual_result = day.solve_part2(problem_input);
        assert_eq!(actual_result, expected_result);
    }
}